src/k2/module/k23si/Module.cpp

```diff
@@ -749,13 +754,17 @@ K23SIPartitionModule::handleRead(dto::K23SIReadRequest&& request, FastDeadline d
         return _makeReadOK(rec);
     }
 
+    count++;
     // record is still pending and isn't from same transaction.
     return _doPush(request.key, versions.WI->data.timestamp, request.mtr, deadline)
-        .then([this, request=std::move(request), deadline](auto&& retryChallenger) mutable {
+        .then([this, request=std::move(request), deadline, count](auto&& retryChallenger) mutable {
             if (!retryChallenger.is2xxOK()) {
                 return RPCResponse(dto::K23SIStatus::AbortConflict("incumbent txn won in read push"), dto::K23SIReadResponse{});
             }
-            return handleRead(std::move(request), deadline);
+            if (count > _config.maxPushCount()) {
+                return RPCResponse(dto::K23SIStatus::InternalError("request has already been pushed, incumbent txn won in read push"), dto::K23SIReadResponse{});
```

Comment:
I don't think this is an internal error. Let's just give back an AbortConflict

Reply:
Why not a retryable error?

Reply:
The transaction should be aborted after this error and it is usually caused by high conflict in transactions, let us use AbortConflict here.

Reply:
changed to AbortConflict